### ISSUE TITLE
Service initializer interface compliance

### ISSUE THREAD

**Issue by sgammon** [grpc/grpc-swift#103]
Using `grpc-swift` latest (as of this writing, `0.3.2`) with _SwiftProtobuf_ latest (`1.0.2`) produces services with the following initializers:

```swift
  /// Create a client that makes insecure connections.
  public required init(address: String) {
    [...]
  }

  /// Create a client that makes secure connections.
  public required init(address: String, certificates: String?, host: String?) {
    [...]
  }
```

In libraries that use multiple services, one might be tempted to create a generic-factory to setup service objects and register them centrally, or otherwise prepare them for use by higher-order code.  Perhaps you want to prepare endpoints in a uniform way, or certs, or what not.
Consider the following example:

```swift
/**
 * Main protocol for a remotely-supported gRPC service.
 */
internal protocol RPCService {
  init(address: String)
  init(address: String, certificates: String?, host: String?)
}

/**
 * Generic RPC service factory.
 */
internal final class RPCServiceFactory<Service: RPCService> {
  static func factory(endpoint: RPCEndpoint) -> Service {
    if let secure = endpoint as? SecureRPCEndpoint {
      // connect over TLS
      return Service.init(
        address: "\(secure.hostname ?? secure.host):\(secure.port)",
        certificates: secure.chain,
        host: secure.hostname)
    } else {
      // connect over plaintext
      return Service.init(
        address: "\(endpoint.host):\(endpoint.port)")
    }
  }
}
```

Whoops! This doesn't work because the initializers in the compiled gRPC code are not designated as `required`. Additionally, they aren't marked as `final`, implying they could be extended, but also aren't marked `open` for cross-module extension, implying they should be extended only inside other compiled code.

I'm working in a Swift package context here, and extending the class from the same module not only seems excessive just to enforce initializer interface compliance, but hard to pull off in a repeatable way if I'm copying Swift into a folder containing custom extension code.

**tl;dr**
Can `grpc-swift` safely add `required` to the initializers for those compiled services, or is there a better approach to generic initialization? I also don't see a compiled/embedded protocol that does this already, perhaps there is and I could use that.

Happy to contribute code if it helps

**Comment by timburks** [comment]
Hi Sam, I need to learn more about Swift initializers!

When you say this:
> This doesn't work because the initializers in the compiled gRPC code are not designated as required.  

Do you mean that your calling code fails to compile? 

I'm confused about which parts of your suggestions are overlapping and which are alternatives. Could you clarify the alternatives? The ones that I think I understand are:

1. Add "required" to the initializers of the generated service. That would fix the compile error that I guess you're seeing in your calling code.

2. Change the access control of the initializers from "public" to "open". That is consistent with my expectations around these methods - I would prefer to allow them to be overridden in any module.

Thanks for raising this!

**Comment by sgammon** [comment]
Yes, sorry, let me clarify :)

My calling code fails to compile. I do believe both options you listed are workable solutions and option 2 would be a fine solution for my use-case.

**Comment by MrMage** [comment]
Hi Sam,

Would this have been fixed by #155? You might be able to use a `Service: ServiceServer` requirement now, I think. Otherwise, what would be left to do — just mark the initializers on `ServiceServer` as `required`?

Thanks,
Daniel

**Comment by sgammon** [comment]
Hey @MrMage, let me update my dependencies and take a look. I'll see if my protocols can extend that one or be eliminated entirely. Thank you! :)

**Comment by cepages** [comment]
I have the same problem trying to create a factory method:

```    
    private mutating func prepareService<T:ServiceClientBase>(typeClass: T.Type) -> T {
        
        let services = self.listOfServices.filter { ( type(of: $0) == typeClass) }
        let instanceServiceFromList = services.count > 0 ? services.first as? T : nil
        if let instanceService = instanceServiceFromList {
            return instanceService
        }
        
        let instanceService = typeClass.init(address: address, secure: false)
        self.listOfServices.append(instanceService)
        return instanceService
    }
```

With this I would able to create any service with:
```
        let service = self.prepareService(typeClass: Ont_Rpc_Instance_InstanceServiceServiceClient.self)
```
 
XCode doesn't compile because the init method I'm using doesn't have the designation `required` so I can't create the class dynamically.

I fixed adding the designation required in the init method:

`required public init(address: String, secure: Bool = true) {`

Obviously this will be deleted when I update this pod in my project, so it would be great if this designation is added to the init methods 

I think this is similar or the same problem that @sgammon was talking about.

**Comment by MrMage** [comment]
@cepages could you please file a pull request? Thanks!

**Comment by cepages** [comment]
Sure, on it :)

**Comment by mushtaque87** [comment]
Took the latest  
 pod 'SwiftGRPC' 
 pod 'SwiftProtobuf'
version using pod update and it got fixed .

Installing BoringSSL 10.0.6 (was 10.0.5 and source changed to `https://github.com/CocoaPods/Specs.git` from `https://github.com/cocoapods/specs.git`)
Installing SwiftGRPC 0.5.1 (was 0.4.3 and source changed to `https://github.com/CocoaPods/Specs.git` from `https://github.com/cocoapods/specs.git`)
Installing gRPC-Core 1.12.0 (was 1.11.0 and source changed to `https://github.com/CocoaPods/Specs.git` from `https://github.com/cocoapods/specs.git`)
Installing Realm 3.7.6 (was 3.7.4 and source changed to `https://github.com/CocoaPods/Specs.git` from `https://github.com/cocoapods/specs.git`)
Installing RealmSwift 3.7.6 (was 3.7.4 and source changed to `https://github.com/CocoaPods/Specs.git` from `https://github.com/cocoapods/specs.git`)

Then all i need was create the session/service with my certificate.

let certPath = Bundle.main.url(forResource: "XYZ", withExtension: "pem")!
        let certificate = try! String(contentsOf: certPath)
        service = Pronounce_PronounceServiceClient(address: Constants.ServerApi.grpcBaseUrl, certificates: certificate)
